### PR TITLE
Add struct arg with optional pre- and postShutdown closers to Serve()

### PR DESCRIPTION
**Pull request by suversky**
Add an argument struct to Serve(), allowing for optional slices of `io.Closers` that can shut down gracefully before and after the server.

-----

This is a precursor for adding queue consumers, so that they can be cleanly shutdown alongside the server.

### CODE REVIEW

File: baseplate.go
Comment (fishy) [inline baseplate.go:132]: style nit: prefer `var be errorsbp.Batch`.

but `server` itself is also a closer, so it's actually easier to just add server into the batch closer, e.g.:

```
bc := batchcloser.New(closers...)
bc.Add(server)
closeChannel <- bc.Close()
```

**Comment by fishy** [comment]
Oh also I just noticed this part:

>This is a precursor for adding queue consumers, so that they can be cleanly shutdown alongside the server.

In that case, the order of close does matter. I think for the default implementation we should close the server before the additional closers, because the server is usually the main drive of all the sub goroutines/tasks. And that behavior should be clearly documented in the function comment doc, so if for some service it makes more sense to close the server the last, they know that they need to handle that manually.

**Comment by pacejackson** [comment]
>  I think for the default implementation we should close the server before the additional closers, because the server is usually the main drive of all the sub goroutines/tasks.

Interesting, I was thinking the _other_ way, but mostly just because the "server" is what drives the healthcheck but I can see it this way too...

**Comment by fishy** [comment]
> > I think for the default implementation we should close the server before the additional closers, because the server is usually the main drive of all the sub goroutines/tasks.
> 
> Interesting, I was thinking the _other_ way, but mostly just because the "server" is what drives the healthcheck but I can see it this way too...

so the case you are talking about is that the queue consumer is the main "server" driving tasks, and the thrift "server" is only for health check? in that case I think we should actually make the queue consumer as the server and thrift server as the additional closer.

here the api doesn't specify what are those closers, so this thing gets muddy fast :)  in our use case we often have thrift server also running kafka producer, so if we close the kafka producer before closing the thrift server, thrift requests can come in after the kafka producer is already closed, and we might try to produce after the producer is closed, and that will cause troubles.

**Comment by pacejackson** [comment]
Part of the point was to avoid writing a specific "queue consumer server", you just use a basic HTTP or Thrift (we use HTTP in python) "server" that only serves healthchecks and include the queue consumer as an extra closer.

**Comment by fishy** [comment]
> Part of the point was to avoid writing a specific "queue consumer server", you just use a basic HTTP or Thrift (we use HTTP in python) "server" that only serves healthchecks and include the queue consumer as an extra closer, but maybe we can't get away from that?

Ok in that case the order is fine,but it does need to be documented

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:90]: In addition to some extra documentation, maybe we should name these closers something more descriptive?  Like "onShutdown" or maybe even "preShutdown"?

**Review by pacejackson — CHANGES_REQUESTED** [review]

**Comment by fishy** [comment]
Actually looking at it more, when people uses `baseplate.Serve` this is the shutdown handler, so if they do have anything that need to be closed after `server.Close` they can't do that. So I can see we likely will need to add something like `PostClosers` in the future.

I'd suggest since we already have breaking changes lined up, it's probably time to change that function's args into a struct, and the closers we add now are `PreClosers`. We don't need to add `PostClosers` now but when we add them in the future that won't be a breaking change.

**Comment by suversky** [comment]
Sounds good to me! I figured the variadic approach would be good for a first stab at it and I'd use the struct approach if you more-knowledgeable folks gave me a reason to :slightly_smiling_face:

File: baseplate.go
Comment (suversky) [inline baseplate.go:88]: I'm not actually sure how formatting (and tabs vs spaces) works in whatever utility is generating API documentation from this, so I'd love feedback on that!

File: baseplate.go
Comment (suversky) [inline baseplate.go:158]: I'm perhaps naively relying on the assumption that `preShutdown` and `postShutdown` will work appropriately if unset (presumably either `nil` or an empty slice, and in either case not problematic for `batchcloser.New` or `batchcloser.Add`.

File: baseplate.go
Comment (suversky) [reply baseplate.go:88]: Interestingly it shows up properly in the `/files` endpoint but improperly in this snippet.

I used spaces when laying out the comment but I guess `go fmt` might have tabified them.

Comment (pacejackson) [reply baseplate.go:88]: I believe tabs make things look like code and yiou should use spaces.  Also, you don't want to do this sort of spacing, i.e.:

```go
// ServeArgs provides a list of arguments to Serve.
// server: Mandatory. Indicates the Server that should be run until
// receiving a shutdown signal.
```

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:92]: 2 things:

1. You actually want to document the fields just above the field itself, not when documenting the struct itself.
2. These need to be capitalized so they are exported, as it is, only things in this package actually have access to these fields (so they could never actually be set by a service).

File: baseplate.go
Comment (pacejackson) [reply baseplate.go:158]: I think this should be fine.

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:156]: style: I would just use `args.preShutdown...` directly here rather than assigning it to a variable that is only used once.

```go
				bc.Add(args.postShutdown...)
```

File: baseplate.go
Comment (pacejackson) [reply baseplate.go:158]: Same style comment as above for this line:

```go
				bc.Add(args.postShutdown...)
```

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:111]: style nit, calling this "serveArgs" is redundant, I think just "args" is simpler.

**Review by pacejackson — CHANGES_REQUESTED** [review]
Can you also add some tests to https://github.com/reddit/baseplate.go/blob/fc4a6864ce01ce5322b65dab8552558dea7278fc/baseplate_test.go#L89 ?

File: baseplate.go
Comment (suversky) [reply baseplate.go:92]: Ooh, I didn't realize that capitalization exporting applied to fields of structs as well as the struct itself. Clever!

File: baseplate.go
Comment (fishy) [inline baseplate.go:101]: super nit: this way the 3 bullet points will be rendered as code block in the godoc site, which works but might look weird. it's slightly better to not put 4 extra spaces before them, and add blank lines between them, e.g.:

```
// other comments
//
// * bullet 1
//
// * bullet 2
```

**Review by fishy — CHANGES_REQUESTED** [review]
You need to fix the test, but otherwise lgtm.

File: baseplate.go
Comment (suversky) [reply baseplate.go:101]: Great nit, and thanks for pointing me to the name `godoc` so I know what to google for now :P

File: httpbp/example_server_test.go
Comment (fishy) [inline httpbp/example_server_test.go:110]: It's a bit weird to define `args` before the `err` check. In general I'd prefer to just inline it in the line it's used, especially in examples. Same for the thrift example.

**Comment by suversky** [comment]
The last thing I've got to do is add some tests for the `PreShutdown` and `PostShutdown` closers, which I'm still trying to reason about.

**Comment by fishy** [comment]
> The last thing I've got to do is add some tests for the `PreShutdown` and `PostShutdown` closers, which I'm still trying to reason about.

We should add that test to make sure that we are closing them in the correct order.

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:83]: You should write comments like a normal sentence so this should be more like:

```suggestion
	// Server is the Server that should be run until receiving a shutdown signal.
	// This is a required argument and baseplate.Serve will panic if this is nil.
```

File: baseplate.go
Comment (pacejackson) [inline baseplate.go:87]: Similar to above:

```suggestion
	// PreShutdown is an optional slice of io.Closers that should be gracefully shut down
	// before the server upon receipt of a shutdown signal.
```

**Review by pacejackson — CHANGES_REQUESTED** [review]

File: baseplate_test.go
Comment (suversky) [inline baseplate_test.go:199]: I cribbed this from `TestServe`, but am not actually too sure if this block (and the provided `args.Server`) are correct for testing a happy shutdown path.

File: baseplate_test.go
Comment (pacejackson) [inline baseplate_test.go:189]: 2 style nits:

1. This should be fine to fit on one line...
2. Modulo that you should just define `args` inline here

```suggestion
				ch <- baseplate.Serve(context.Background(), baseplate.ServeArgs{
					Server:       newWaitServer(t, bp, time.Millisecond),
					PreShutdown:  []io.Closer{pre},
					PostShutdown: []io.Closer{post},
				})
```

File: baseplate_test.go
Comment (pacejackson) [inline baseplate_test.go:181]: This probably doesn't need to be it's own subtest, it's just this one test and I don't think you really need to add more.

File: baseplate_test.go
Comment (pacejackson) [reply baseplate_test.go:199]: Seems fine to me, do you have any specific concerns?

**Review by pacejackson — APPROVED** [review]
LGTM modulo a couple style updates in the new test.

File: baseplate_test.go
Comment (fishy) [reply baseplate_test.go:199]: `err` is assigned but never used in its current state? Do you mean to check it? if not you can just do `<-ch` without assigning it to anything.

File: baseplate_test.go
Comment (fishy) [inline baseplate_test.go:202]: nit: for ints (`len()`) you don't really need `%#v`. just `%v` or `%d` will do.

File: baseplate_test.go
Comment (fishy) [inline baseplate_test.go:210]: for `time.Time` types, `%v` will work better than `%#v`.

File: baseplate_test.go
Comment (fishy) [inline baseplate_test.go:201]: shouldn't this be `!=` instead of `>`? you want to fail it if it's not called as well.

**Comment by suversky** [comment]
Thanks so much for all the feedback!

File: baseplate_test.go
Comment (fishy) [inline baseplate_test.go:189]: If we failed to get the pid, the already started goroutine will just be leaked (until all tests are done). I'd prefer to get pid first, fatal if failed, then start goroutine, sleep, and signal.

File: baseplate_test.go
Comment (fishy) [inline baseplate_test.go:203]: I'd prefer not to overly use Fatal over Error in tests. The previous ones are fine: We can't continue the test if we can't trigger graceful shutdown, getting `pref.ts[0]` could panic if `len(pre.ts) < 1`, but here this is just an error, not some thing fatal.

**Review by fishy — APPROVED** [review]
The two remaining comments are mostly nitpicks and not blocking.

File: baseplate_test.go
Comment (fishy) [reply baseplate_test.go]: 🔕 We probably should check `err != nil` though?

**Review by fishy — APPROVED** [review]

**Pull request merged**
